### PR TITLE
507 backup contingency

### PR DESCRIPTION
**Pull request by Evie-Skinner18**
## Context

- Rewrite of the backup contingency story
- Using Spring Batch to do this like the Export feature already does
- Performance gains from doing it this way

## Guidance to review

- there's a lot of duplication
- How can I improve? E.g I wanted to try and make it polymorphic by having both -SpreadsheetRows implement a SpreadsheetRow interface

## Link to Trello card

https://trello.com/c/lTYLhdgC/507-backup-contingency-as-a-sar-mcc-operator-during-a-business-continuity-event-i-can-access-a-recent-backup-of-everything-else-in-t

### CODE REVIEW

File: backoffice/src/gateways/exports/IExportsGateway.ts [hunk 1]
```diff
@@ -19,5 +19,6 @@ export interface IExportsGateway {
   searchExportData(
     searchForm: ExportSearchFormProps
   ): Promise<IBeaconExportSearchResult>;
-  getBackupExportFile(): Promise<AxiosResponse>;
+  downloadExistingBackupFile(): Promise<AxiosResponse>;
```
Comment (slkendell-mt) [inline backoffice/src/gateways/exports/IExportsGateway.ts:22]: Why is this added twice?

File: service/src/main/java/uk/gov/mca/beacons/api/export/ExportFileNamer.java [hunk 2]
```diff
@@ -51,17 +51,36 @@ public ExportFileNamer(Clock clock) {
...
       "-" +
       protectiveMarking +
       fileType.extension
     );
   }
 
+  private String getFilename(
+    BeaconsDataWorkbookRepository.OperationType operationType
+  ) {
+    String filename = "Beacons_Data";
```
Comment (slkendell-mt) [inline service/src/main/java/uk/gov/mca/beacons/api/export/ExportFileNamer.java:71]: If this is all the function does, I'd probably switch it out to:
```
switch (operationType) {
      case EXPORT:
        return "Beacons_Export_Data";
      case BACKUP:
        return "Beacons_Backup_Data";
      default:
        return "Beacons_Data";
    }

File: service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupBeaconToSpreadsheetRowItemProcessor.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import uk.gov.mca.beacons.api.note.domain.Note;
+import uk.gov.mca.beacons.api.registration.application.RegistrationService;
+import uk.gov.mca.beacons.api.registration.domain.Registration;
+
+@Component
+class BackupBeaconToSpreadsheetRowItemProcessor
+  implements ItemProcessor<Beacon, BackupSpreadsheetRow> {
+
+  private final RegistrationService registrationService;
+  //  private final BeaconOwnerReadOnlyRepository beaconOwnerRepository;
```
Comment (slkendell-mt) [inline service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupBeaconToSpreadsheetRowItemProcessor.java:32]: Remove commented out

File: service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupBeaconToSpreadsheetRowItemProcessor.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  private final ExportMapper exportMapper;
+
+  private static final DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern(
+    "dd-MM-yyyy"
+  );
+
+  @Autowired
+  public BackupBeaconToSpreadsheetRowItemProcessor(
+    RegistrationService registrationService,
+    //BeaconOwnerReadOnlyRepository beaconOwnerRepository,
```
Comment (slkendell-mt) [inline service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupBeaconToSpreadsheetRowItemProcessor.java:47]: Remove commented out

File: service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupBeaconToSpreadsheetRowItemProcessor.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    //          BeaconUseReadOnlyRepository beaconUseRepository,
+    //          EmergencyContactReadOnlyRepository emergencyContactRepository,
+    NoteService noteService,
+    AccountHolderService accountHolderService,
+    ExportMapper exportMapper
+  ) {
+    this.registrationService = registrationService;
+    //    this.beaconOwnerRepository = beaconOwnerRepository;
```
Comment (slkendell-mt) [inline service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupBeaconToSpreadsheetRowItemProcessor.java:55]: Likewise

File: service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupScheduler.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  @Autowired
+  public BackupScheduler(BackupXlsxExporter backupXlsxExporter) {
+    this.backupXlsxExporter = backupXlsxExporter;
+  }
+
+  @Scheduled(cron = "0 0 2 * * MON") // Every Monday at 2am
+  @SchedulerLock(
+    name = "xlsxBackup_scheduledTask",
+    lockAtLeastFor = "15m", // 15 minutes
+    lockAtMostFor = "30m" // 30 minutes
```
Comment (slkendell-mt) [inline service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupScheduler.java:23]: Is 30m long enough ?

File: service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupToXlsxJobManager.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * Initiates export jobs using Spring Batch.
+ */
+@Component
+@Slf4j
+public class BackupToXlsxJobManager {
+
+  private final JobLauncher jobLauncher;
+
+  // todo: how do I reuse one XlsxJobManager and decide at runtime which job I want to run
```
Comment (slkendell-mt) [inline service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupToXlsxJobManager.java:27]: Do you still need this comment?

Comment (Evie-Skinner18) [reply service/src/main/java/uk/gov/mca/beacons/api/export/xlsx/backup/BackupToXlsxJobManager.java:27]: If poss I would like to know how to address the question in the comment but maybe we should save refactoring this feature for a future iteration
